### DISCUSSION TITLE
Including NXP HAL modules in network driver

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#53181] [Q&A] [answered]
Hey, 

I'm currently rewriting NXP MIMXRT1060_EVKB board's network driver (eth_mcux.c) for our custom board which based on the same microcontroller, and I'm finding it hard to include and use ENET_* functions in my zephyr module driver. 

In my driver I tried to include the needed header files:
#include "fsl_enet_mdio.h"
#include "fsl_phy.h"

and in my CMakeLists.txt file I included the appropriate directories:
    zephyr_include_directories(${ZEPHYR_HAL_NXP_MODULE_DIR}/mcux/mcux-sdk/drivers/enet)
    zephyr_include_directories(${ZEPHYR_HAL_NXP_MODULE_DIR}/mcux/mcux-sdk/components/phy)
    zephyr_include_directories(${ZEPHYR_HAL_NXP_MODULE_DIR}/mcux/mcux-sdk/components/phy/mdio/enet)

but it doesn't work for me.

Did someone deal with that problem ?

Thanks in advance, 
Ofir.

**Comment by dbaluta** [comment]
It might be helpful to also post the errors you get.

**Reply by ofirshe** [reply]
Thanks Daniel, 

here is the compilation output: 
[307/317] Linking C executable zephyr\zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map C:/Users/ofirs/epu-bsp/build/zephyr/zephyr_pre0.map
cmd.exe /C "cd . && C:\Users\ofirs\epu-bsp\gcc-arm-none-eabi-10.3-2021.10\bin\arm-none-eabi-gcc.exe  --specs=nosys.specs   -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr\zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=C:/Users/ofirs/epu-bsp/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/soc/arm/common/cortex_m/libsoc__arm__common__cortex_m.a  zephyr/subsys/net/libsubsys__net.a  zephyr/subsys/net/l2/ethernet/libsubsys__net__l2__ethernet.a  zephyr/subsys/net/ip/libsubsys__net__ip.a  zephyr/subsys/net/lib/config/libsubsys__net__lib__config.a  zephyr/subsys/random/libsubsys__random.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/i2c/libdrivers__i2c.a  zephyr/drivers/spi/libdrivers__spi.a  zephyr/drivers/memc/libdrivers__memc.a  zephyr/drivers/flash/libdrivers__flash.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/entropy/libdrivers__entropy.a  zephyr/drivers/timer/libdrivers__timer.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  modules/hal_nxp/hal_nxp/lib..__modules__hal__nxp.a  modules/mbedtls/libmodules__mbedtls.a  modules/pcf2131/lib..__src__drivers__rtc__pcf2131.a  modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/thumb/v7e-m/nofp"  -LC:/Users/ofirs/epu-bsp/build/zephyr  -lgcc  zephyr/libcode_relocation_source_lib.a  zephyr/arch/common/libisr_tables.a  -no-pie  -mcpu=cortex-m7  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cmd.exe /C "cd /D C:\Users\ofirs\epu-bsp\build\zephyr && "C:\Program Files\CMake\bin\cmake.exe" -E echo ""
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a(kinetis_ethernet_xsight.c.obj): in function `kinetis_ethernet_xsight_common_isr':
C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:518: undefined reference to `ENET_ReceiveIRQHandler'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:524: undefined reference to `ENET_TransmitIRQHandler'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a(kinetis_ethernet_xsight.c.obj): in function `net_if_mcast_cb':
C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:424: undefined reference to `ENET_AddMulticastGroup'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:426: undefined reference to `ENET_LeaveMulticastGroup'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a(kinetis_ethernet_xsight.c.obj): in function `kinetis_ethernet_xsight_set_config':
C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:484: undefined reference to `ENET_SetMacAddr'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a(kinetis_ethernet_xsight.c.obj): in function `kinetis_ethernet_xsight_init':
C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:330: undefined reference to `ENET_GetDefaultConfig'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:355: undefined reference to `ENET_Init'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:362: undefined reference to `ENET_SetSMI'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:364: undefined reference to `ENET_SetCallback'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a(kinetis_ethernet_xsight.c.obj): in function `eth_init':
C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:406: undefined reference to `enet_ops'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a(kinetis_ethernet_xsight.c.obj): in function `eth_tx':
C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:170: undefined reference to `ENET_SendFrame'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:177: undefined reference to `ENET_ReclaimTxDescriptor'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: modules/kinetis_ethernet_xsight/lib..__src__drivers__ethernet__kinetis_ethernet_xsight.a(kinetis_ethernet_xsight.c.obj): in function `eth_rx':
C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:196: undefined reference to `ENET_GetRxFrameSize'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:205: undefined reference to `ENET_GetRxErrBeforeReadFrame'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:258: undefined reference to `ENET_ReadFrame'
c:/users/ofirs/epu-bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: C:/Users/ofirs/epu-bsp/src/drivers/ethernet/kinetis_ethernet_xsight/kinetis_ethernet_xsight.c:227: undefined reference to `ENET_ReadFrame'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\ofirs\epu-bsp\build'

**Reply by ofirshe** [reply]
Do you have any suggestions ? @dbaluta

**Comment by DerekSnell** [comment, marked as answer]
Hi @ofirshe ,
From the errors, it appears your build is not including `fsl_enet.c` in the NXP HAL.  This driver is included in CMake in [hal_nxp.cmake](https://github.com/zephyrproject-rtos/hal_nxp/blob/0ee50e0796d66307c8a99e9178140e7432b1eaba/mcux/hal_nxp.cmake#L96).  

You show using `zephyr_include_directories()` to include the enet folder.  My understanding is this Cmake command tells the compiler where the include files are, like .h files.  But a command like `zephyr_library_sources()` is needed to include a source file in the build, like here in [hal_nxp.cmake](  zephyr_library_sources(mcux-sdk/components/phy/device/phyksz8081/fsl_phyksz8081.c)).  Or here is how [driver_enet.cmake](https://github.com/zephyrproject-rtos/hal_nxp/blob/0ee50e0796d66307c8a99e9178140e7432b1eaba/mcux/mcux-sdk/drivers/enet/driver_enet.cmake#L6) adds `fsl_enet.c` to the build.

BTW, do you also have `CONFIG_ETH_MCUX` selected for your board?

Best regards

**Reply by ofirshe** [reply]
Thanks @DerekSnell, 

I added 
zephyr_library_sources(${ZEPHYR_HAL_NXP_MODULE_DIR}/mcux/mcux-sdk/drivers/enet/fsl_enet.c)
zephyr_library_sources(${ZEPHYR_HAL_NXP_MODULE_DIR}/mcux/mcux-sdk/components/phy/mdio/enet/fsl_enet_mdio.c)

to my CMakeLists.txt and now it can be compiled. 

For your question, I don't have CONFIG_ETH_MCUX selected on my board. I created a new one CONFIG_KINETIS_ETHERNET_XSIGHT for my custom driver and enabled it in the project configuration file. 

now the compilation is succeeded but zephyr doesn't recognize as network driver:

CMake Warning at zephyr/CMakeLists.txt:824 (message):
  No SOURCES given to Zephyr library: drivers__ethernet

  Excluding target from build.


-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/ofirs/epu-bsp/build
-- west build: building application
[9/19] Linking C executable zephyr\zephyr_pre0.elf

[13/19] Linking C executable zephyr\zephyr_pre1.elf

[19/19] Linking C executable zephyr\zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      133596 B         8 MB      1.59%
             RAM:      129808 B       768 KB     16.51%
            ITCM:        3520 B       128 KB      2.69%
            DTCM:       13908 B       128 KB     10.61%
        IDT_LIST:          0 GB         2 KB      0.00%

Do you know why ?

**Reply by ofirshe** [reply]
I saw that the driver is being added by:
include_driver_ifdef(CONFIG_ETH_MCUX			enet		driver_enet)

In my case, I create all my drivers as zephyr modules and just add it in the CmakeLists.txt file like that: 
list(APPEND ZEPHYR_EXTRA_MODULES
  ${CMAKE_CURRENT_SOURCE_DIR}/src/drivers/ethernet/kinetis_ethernet_xsight
)

but it seems not to be sufficient because zephyr doesn't recognize it as ethernet driver. Also enabled it in the dts: 
&enet {
	status = "okay";
	compatible = "nxp,kinetis-ethernet-xsight";
	pinctrl-0 = <&pinmux_enet>;
	pinctrl-names = "default";
	int-gpios = <&gpio1 10 GPIO_ACTIVE_HIGH>;
	reset-gpios = <&gpio1 9 GPIO_ACTIVE_HIGH>;
};


Do you have any suggestions @DerekSnell ?

Thanks,
Ofir.

**Comment by DerekSnell** [comment]
Hi @ofirshe ,
From your build log, it appears your application builds successfully.  The driver files you need must all be included in the build, or the linker would fail.  Perhaps the CMake warning is just an artifact of how you are managing the CMake files for your board.

Are you aware of an actual error, besides this CMake warning?

**Reply by ofirshe** [reply]
@DerekSnell  
No, 

I don't have any warning but this one. 
C:\Users\ofirs\epu-bsp>west build -b xsightlabs_evb_a -- -DE1_STUB=1
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base (cached)).
-- Application: C:/Users/ofirs/epu-bsp
-- Cache files will be written to: C:/Users/ofirs/epu-bsp/zephyr/.cache
-- Zephyr version: 3.2.99 (C:/Users/ofirs/epu-bsp/zephyr)
-- Found west (found suitable version "0.13.1", minimum required is "0.7.1")
-- Board: xsightlabs_evb_a
-- Found toolchain: gnuarmemb (C:\Users\ofirs\epu-bsp\gcc-arm-none-eabi-10.3-2021.10)
-- Found BOARD.dts: C:/Users/ofirs/epu-bsp/boards/arm/xsightlabs_evb/xsightlabs_evb_a.dts
node '/soc/i2c@403f8000/fs1403@8' compatible 'tdk,fs140x' has unknown vendor prefix 'tdk'
-- Generated zephyr.dts: C:/Users/ofirs/epu-bsp/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: C:/Users/ofirs/epu-bsp/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: C:/Users/ofirs/epu-bsp/build/zephyr/dts.cmake
Parsing C:/Users/ofirs/epu-bsp/zephyr/Kconfig
Loaded configuration 'C:/Users/ofirs/epu-bsp/build/zephyr/.config'
No change to configuration in 'C:/Users/ofirs/epu-bsp/build/zephyr/.config'
No change to Kconfig header in 'C:/Users/ofirs/epu-bsp/build/zephyr/include/generated/autoconf.h'
Load components for MIMXRT1062:
driver_common component is included.
driver_reset component is included.
device_CMSIS component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_trng component is included.
driver_lpi2c component is included.
driver_lpspi component is included.
driver_lpuart component is included.
driver_igpio component is included.
driver_flexspi component is included.
driver_cache_armv7_m7 component is included.
CMake Warning at zephyr/CMakeLists.txt:824 (message):
  No SOURCES given to Zephyr library: drivers__ethernet

  Excluding target from build.


-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/ofirs/epu-bsp/build
-- west build: building application
[4/12] Linking C executable zephyr\zephyr_pre0.elf

[7/12] Linking C executable zephyr\zephyr_pre1.elf

[12/12] Linking C executable zephyr\zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      133596 B         8 MB      1.59%
             RAM:      129808 B       768 KB     16.51%
            ITCM:        3520 B       128 KB      2.69%
            DTCM:       13908 B       128 KB     10.61%
        IDT_LIST:          0 GB         2 KB      0.00%


I believe that this warning occurs because now CONFIG_ETH_MCUX is not enabled and this line include_driver_ifdef(CONFIG_ETH_MCUX			enet		driver_enet) not oocured. How can I declare on my drive as 'enet' driver ? 

Thanks

**Reply by ofirshe** [reply]
ok because this issue was resolved, I'll open a new one regarding the new problem.
